### PR TITLE
Search stack prototype

### PR DESCRIPTION
**Pull request by fkling**
Depends on #28491 

See screenshots and video. I decided to not go with a sidebar for the prototype because that would impact the layout of a lot of pages, and that wasn't worth it for the prototype IMO.

Instead I have a small sticky box in the lower right corner. It collapses to a single icon to not cover too much content (e.g. the other information on blob pages).

To enable this feature, set `enableSearchStack` in the experimental features settings.

**Note:**
- I'm currently not storing line ranges for the file entries, it doesn't look like I have access to those where I'm adding the entry (`BlobPage`).
- File blocks don't render yet in notebooks

https://user-images.githubusercontent.com/179026/144625664-02510656-c192-4e4e-ac3c-8954c9d27746.mp4

### CODE REVIEW

**Comment by marisak** [comment]
It's so cool to see this come to life! I'm excited to start research on this; I think this prototype will enable a lot of really fruitful conversations. 

I noticed that longer queries don't wrap very well in the stacks bar. (Stacks bar? Widget? Popover?) Is there a particular reason for choosing 15rem as the max width? I think it could be a bit wider, which would help with some of the wrapping. I'm also curious if it's possible to wrap at a character count; this would make some of the queries hard to read, but that's better than overflowing out of the container. Screenshot below: 
screenshot of PR 28552 showing text overflow

File: client/web/src/search/SearchStack.tsx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+import { SyntaxHighlightedSearchQuery } from '../components/SyntaxHighlightedSearchQuery'
+import { PageRoutes } from '../routes.constants'
+import { useExperimentalFeatures } from '../stores'
+import { SearchEntry, FileEntry, useSearchStackState, restorePreviousSession } from '../stores/searchStack'
+
+import { BlockInput } from './notebook'
+import { serializeBlocks } from './notebook/helpers'
+import styles from './SearchStack.module.scss'
+
+export const SearchStack: React.FunctionComponent = () => {
```
Comment (limitedmage) [inline client/web/src/search/SearchStack.tsx:24]: Can some tests be added for this component? I think both visual and unit tests could be beneficial.

File: client/web/src/stores/searchStack.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
+import { useEffect } from 'react'
+import create from 'zustand'
+
+import { SearchPatternType } from '@sourcegraph/shared/src/graphql-operations'
+import { FilterType } from '@sourcegraph/shared/src/search/query/filters'
+import { omitFilter } from '@sourcegraph/shared/src/search/query/transformer'
+import { FilterKind, findFilter } from '@sourcegraph/shared/src/search/query/validate'
+
+import { useExperimentalFeatures } from './experimentalFeatures'
+
```
Comment (limitedmage) [inline client/web/src/stores/searchStack.ts:10]: Is it possible to add unit tests for these hooks?

**Review by limitedmage — APPROVED** [review]
This looks great! One thing I immediately noticed is that searches and files are being grouped, so it's not really a single "stack". I think a chronological view that includes both searches and files might be more true to the idea of a stack. I think this order should also be conserved when creating the notebook. It would also be nice to have a button to clear the stack or remove individual items.

At this point, I think this is ready to be checked in (once some tests are added) under the experimental feature flag and we can keep improving on it.

File: client/web/src/search/notebook/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { serializeBlockInput } from './serialize'
+
+import { BlockInput } from '.'
+
+export function serializeBlocks(blocks: BlockInput[]): string {
```
Comment (novoselrok) [inline client/web/src/search/notebook/helpers.ts:5]: You can also put this into `notebook/serialize.ts` file.

File: client/web/src/repo/blob/BlobPage.tsx [hunk 4]
```diff
@@ -79,6 +79,18 @@ export const BlobPage: React.FunctionComponent<Props> = props => {
         props.telemetryService.logViewEvent('Blob', { repoName, filePath })
     }, [repoName, commitID, filePath, renderMode, props.telemetryService])
 
+    useSearchStack(
+        useMemo(
+            () => ({
+                type: 'file',
+                path: filePath,
+                repo: repoName,
+                revision,
```
Comment (novoselrok) [inline client/web/src/repo/blob/BlobPage.tsx:88]: To access the line range here, you can use extract the `parseQueryAndHash` function call that is already used below (https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/repo/blob/BlobPage.tsx?L188). It returns the line information from the URL. You can move the result to a separate variable and then use it in both places.

**Review by novoselrok — APPROVED** [review]
Similar to Juliana, I expected a chronological view that includes both searches and files. And the order would be kept when converted to a notebook. 

I think grouping by type could be provided as an optional action, e.g. when you click on a button. But for the initial view I would expect a simple list.

**Comment by fkling** [comment]
@marisak

> I noticed that longer queries don't wrap very well in the stacks bar. (Stacks bar? Widget? Popover?) Is there a particular reason for choosing 15rem as the max width? 


Thank you for catching that, I didn't test that 😬 (nor did I test the max height). I used 15rem when I originally implemented this as a side bar, that seemed to be a good width to not make things too crammed on the page. But that's less of a concern with the the popover (widget? 🤷🏻‍♂️ )

> I think it could be a bit wider, which would help with some of the wrapping. I'm also curious if it's possible to wrap at a character count; this would make some of the queries hard to read, but that's better than overflowing out of the container.

Yes, I'll make it wider. Probably can't wrap on character count but we can certainly make it wrap.

---

@novoselrok, @limitedmage 

>  I think a chronological view that includes both searches and files might be more true to the idea of a stack
> Similar to Juliana, I expected a chronological view that includes both searches and files. And the order would be kept when converted to a notebook.

Thank you for this comments! I'm sorry, I should have added my reasoning to the description (that's what you get for finishing things up on a Friday): I might be overthinking this but my initial impression was that implementing a stack that works "well" can be quite tricky (though I can be convinced that it's not).

- If we implement it as a true stack and pop things off when you go back to a previous result, then it's good for navigating (there is a real sense of going back to a certain point and starting something new from that point), but not great for creating notebooks because you loose information about the "previous paths".
- If we just add things chronologically, i.e. always going forward and never truly "go back", then it might be less clear where exactly you are in the "timeline". However, now that I'm writing this I feel like this might actually be an OK tradeoff.

I do want a proper stack eventually but felt like we need more time to nail down how it should work. So in the meantime (the first version) would avoid this problem entirely be removing the timeline/stack aspect entirely and just give you access to the entries (ideally without raising any expectations in the user about "order").
But I also acknowledge that it's not ideal for notebook creation.

So if you (and others (cc @benvenker, @lguychard)) feel strongly about this, I can implement it as a more traditional timeline (a sense of "where you are" could be achieved buy highlighting the entry that matches your current search or file).

Let me know what you think?

File: client/web/src/stores/searchStack.ts (hunk 2 above)
Comment (fkling) [reply client/web/src/stores/searchStack.ts:10]: That is from the other PR. I'll see if we already have tests for settings which I can reuse.

File: client/web/src/search/SearchStack.tsx (hunk 1 above)
Comment (fkling) [reply client/web/src/search/SearchStack.tsx:24]: I didn't want to spend a ton of time writing tests for a prototype, but I can put together some storybook tests.

**Comment by benvenker** [comment]
@fkling I like what you have but think the timeline view is also valuable and probably expected. You and I had talked about possibly combining those aspects, similar to what VS Code does with the timeline and outline panes in the left panel.

**Comment by lguychard** [comment]
I also think going with a simple list of chronologically-added entries makes the least assumptions for a first iteration.

**Comment by fkling** [comment]
Updated UI and behavior: 

https://user-images.githubusercontent.com/179026/145242807-9d4949ca-3c5b-4095-a6be-587660bfc136.mp4

**Comment by benvenker** [comment]
Looks great @fkling, ready to ship it?

**Comment by fkling** [comment]
@benvenker Yep, CI is finally green (or at least as green as it can get)

**Pull request merged**